[DOCS-6513] Add OP datadog archives to setup docs (#20541)
DataDog/documentation: content/en/observability_pipelines/setup/datadog_with_archiving.md

---

> title: Set Up Datadog Archiving with Observability Pipelines

Reviewer: Let's do: "Set up Observability Pipelines to ship logs to S3 in Datadog rehydratable format and to Datadog"

kind: Documentation
further_reading:
  - link: "/observability_pipelines/production_deployment_overview/"
    tag: "Documentation"
    text: "Production deployment design and principles for the Observability Pipelines Worker"
  - link: "https://dtdg.co/d22op"
    tag: "Learning Center"
    text: "Safe and Secure Local Processing with Observability Pipelines"
---

{{< site-region region="gov" >}}
<div class="alert alert-warning">Observability Pipelines is not available on the US1-FED Datadog site.</div>
{{< /site-region >}}

## Overview

The [Observability Pipelines Worker][1] can collect, process, and route logs and metrics from any source to any destination. Using Datadog, you can build and manage all of your Observability Pipelines Worker deployments at scale.

This guide walks you through deploying the Worker in your common tools cluster and configuring it to send logs to a cloud storage for archiving.

## Assumptions
* You are already using Datadog and want to use Observability Pipelines.
* You have administrative access to the clusters where the Observability Pipelines Worker is going to be deployed, as well as to the workloads that are going to be aggregated.

> * You have a common tools or security cluster for your environment to which all other clusters are connected.

Author: I think this is supposed to say "You have a common tools cluster or a security cluster".

> * You have a common tools or security cluster for your environment to which all other clusters are connected.

Author: updated

{{% /tab %}}
{{% tab "AWS EKS" %}}

{{% op-datadog-archives-s3-setup %}}

3. [Create a service account][1] to use the policy you created above.

[1]: https://docs.aws.amazon.com/eks/latest/userguide/associate-service-account-role.html

{{% /tab %}}
{{% tab "APT-based Linux" %}}

{{% op-datadog-archives-s3-setup %}}

3. Create an IAM user and attach the above policy to it. Create access credentials for the IAM user. Save these credentials as `AWS_ACCESS_KEY` and `AWS_SECRET_ACCESS_KEY`.

{{% /tab %}}
{{% tab "RPM-based Linux" %}}

{{% op-datadog-archives-s3-setup %}}

3. Create an IAM user and attach the policy above to it. Create access credentials for the IAM user. Save these credentials as `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.

{{% /tab %}}
{{% tab "Terraform (AWS)" %}}

{{% op-datadog-archives-s3-setup %}}

3. Attach the policy to the IAM Instance Profile that is created with Terraform, which you can find under the `iam-role-name` output.

{{% /tab %}}
{{< /tabs >}}

### Connect the S3 bucket to Datadog Log Archives

> You need to connect the S3 bucket you created earlier to Datadog Log Archives so that you can rehydrate the archives later on.

Author: Laving as is because I think flow-wise it makes sense to explain what they have to do with the S3 bucket they just created.

1. Navigate to Datadog [Log Forwarding][5].

> 1. Click **Add a new archive**.

Author: It should be `+Add Archive`, will update it.

1. Add a query that filters out all logs going through log pipelines so that those logs do not go into this archive. For example, add the query `observability_pipelines_read_only_archive`, assuming that no logs going through the pipeline have that tag added.
1. Select **AWS S3**.
1. Select the AWS Account that your bucket is in.
1. Enter the name of the S3 bucket.
1. Optionally, enter a path.
1. Check the confirmation statement.
1. Optionally, add tags and define the maximum scan size for rehydration. See [Advanced settings][6] for more information.
1. Click **Save**.

See the [Log Archives documentation][7] for additional information.

## Install the Observability Pipelines Worker

{{< tabs >}}
{{% tab "Docker" %}}

The Observability Pipelines Worker Docker image is published to Docker Hub [here][1].

1. Download the [sample pipeline configuration file][2].

2. Run the following command to start the Observability Pipelines Worker with Docker:

    ```shell
    docker run -i -e DD_API_KEY=<API_KEY> \
      -e DD_OP_PIPELINE_ID=<PIPELINE_ID> \
      -e DD_SITE=<SITE> \
      -e AWS_ACCESS_KEY_ID=<AWS_ACCESS_KEY_ID> \
      -e AWS_SECRET_ACCESS_KEY=<AWS_SECRET_ACCESS_KEY> \

>       -p 8282:8282 \

Reviewer: Add to the command
```
DD_ARCHIVES_BUCKET=<AWS_BUCKET_NAME>  
DD_ARCHIVES_SERVICE_ACCOUNT=<BUCKET_AWS_REGION>
```

and on line 188, add the following info about these two values, as we do in [this screenshot](https://a.cl.ly/5zubknyq)

Author: updated

      -v ./pipeline.yaml:/etc/observability-pipelines-worker/pipeline.yaml:ro \
      datadog/observability-pipelines-worker run
    ```
    
    Replace `<API_KEY>` with your Datadog API key, `<PIPELINES_ID>` with your Observability Pipelines configuration ID, and `<SITE>` with {{< region-param key="dd_site" code="true" >}}. Replace `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` with the AWS credentials you created earlier. `./pipeline.yaml` must be the relative or absolute path to the configuration you downloaded in Step 1.
  
[1]: https://hub.docker.com/r/datadog/observability-pipelines-worker
[2]: /resources/yaml/observability_pipelines/archives/pipeline.yaml
{{% /tab %}}
{{% tab "AWS EKS" %}}
1. Download the [Helm chart][1] for AWS EKS.

> 2. In the Helm chart, replace the `datadog.apiKey` and `datadog.pipelineId` values to match your pipeline and use {{< region-param key="dd_site" code="true" >}} for the `site` value. Replace `${DD_ARCHIVES_SERVICE_ACCOUNT}` in the Helm config with the service account you created earlier. Then, install it in your cluster with the following commands:

Reviewer: Could you add the following sentences in [this screenshot](https://a.cl.ly/KouN5YB0)? These values also need to be replaced in the helm chart

Also, let's turn the last sentence that starts with `then` into step 3

Reviewer: Also linking you the [Figma design](https://www.figma.com/file/G6kPDXDo6aaBQvjaDpKpXG/Use-Case-Informed-Onboarding?type=design&node-id=1597-39264&mode=design&t=Jp6Mi0KFxHRBLhtl-0) with the copies

    ```shell
    helm repo add datadog https://helm.datadoghq.com
    ```
    ```shell
    helm repo update
    ```
    ```shell
    helm upgrade --install \
        opw datadog/observability-pipelines-worker \
        -f aws_eks.yaml
    ```

[1]: /resources/yaml/observability_pipelines/archives/aws_eks.yaml
{{% /tab %}}

{{% tab "APT-based Linux" %}}
1. Run the following commands to set up APT to download through HTTPS:

>     ```

Reviewer 2: Non-blocker: are sudo command also shell commands? ```shell

Author: I'll add `shell` in the next iteration because I'll need to separate out all of the commands.

    curl https://keys.datadoghq.com/DATADOG_APT_KEY_F14F620E.public | sudo gpg --no-default-keyring --keyring /usr/share/keyrings/datadog-archive-keyring.gpg --import --batch
    curl https://keys.datadoghq.com/DATADOG_APT_KEY_C0962C7D.public | sudo gpg --no-default-keyring --keyring /usr/share/keyrings/datadog-archive-keyring.gpg --import --batch
    ```

3. Run the following commands to update your local `apt` repo and install the Worker:

    ```
    sudo apt-get update
    sudo apt-get install observability-pipelines-worker datadog-signing-keys
    ```

4. Add your keys and the site ({{< region-param key="dd_site" code="true" >}}) to the Worker's environment variables:

    ```
    sudo cat <<-EOF > /etc/default/observability-pipelines-worker
    DD_API_KEY=<API_KEY>
    DD_OP_PIPELINE_ID=<PIPELINE_ID>
    DD_SITE=<SITE>
    AWS_ACCESS_KEY_ID=<AWS_ACCESS_KEY_ID>
    AWS_SECRET_ACCESS_KEY=<AWS_SECRET_ACCESS_KEY>
    EOF
    ```

5. Download the [sample configuration file][1] to `/etc/observability-pipelines-worker/pipeline.yaml` on the host.

> 

Author: Should we mention here to update the sample config with the archives bucket and region information? https://github.com/DataDog/documentation/pull/20541/files#diff-7c35f7e1358cf5f87d9f00b6526f58ea5959c488b719387b280f894d5

Reviewer: The command on 242 to 248 should be updated, actually ([example](https://a.cl.ly/NQu5mxO9))

So I'd add right under line 247
```
DD_ARCHIVES_BUCKET=<AWS_BUCKET_NAME> 
DD_ARCHIVES_SERVICE_ACCOUNT=<BUCKET_AWS_REGION>
```

    ```
    cat <<EOF > /etc/yum.repos.d/datadog-observability-pipelines-worker.repo
    [observability-pipelines-worker]
    name = Observability Pipelines Worker
    baseurl = https://yum.datadoghq.com/stable/observability-pipelines-worker-1/\$basearch/
    enabled=1
    gpgcheck=1
    repo_gpgcheck=1
    gpgkey=https://keys.datadoghq.com/DATADOG_RPM_KEY_CURRENT.public
           https://keys.datadoghq.com/DATADOG_RPM_KEY_B01082D3.public
           https://keys.datadoghq.com/DATADOG_RPM_KEY_FD4BF915.public
           https://keys.datadoghq.com/DATADOG_RPM_KEY_E09422B3.public
    EOF
    ```

   **Note:** If you are running RHEL 8.1 or CentOS 8.1, use `repo_gpgcheck=0` instead of `repo_gpgcheck=1` in the configuration above.

2. Update your packages and install the Worker:

    ```
    sudo yum makecache
    sudo yum install observability-pipelines-worker
    ```

3. Add your keys and the site ({{< region-param key="dd_site" code="true" >}}) to the Worker's environment variables:

    ```
    sudo cat <<-EOF > /etc/default/observability-pipelines-worker
    DD_API_KEY=<API_KEY>
    DD_OP_PIPELINE_ID=<PIPELINE_ID>
    DD_SITE=<SITE>
    AWS_ACCESS_KEY_ID=<AWS_ACCESS_KEY_ID>

>     AWS_SECRET_ACCESS_KEY=<AWS_SECRET_ACCESS_KEY>

Reviewer: Same here, let's add command and info about bucket and region: https://a.cl.ly/kpuRr4NJ

    default_api_key: "$${DD_API_KEY}"
    compression: gzip
    buffer:
       type: disk
       max_size: 257698037760
  datadog_metrics:
    type: datadog_metrics
    inputs:
      - metrics_add_dd_tags
    default_api_key: "$${DD_API_KEY}"
    buffer:
      type: disk
      max_size: 51539607552
  ## This sink writes logs to an S3 bucket of your choice, in a format that Datadog
  ## can rehydrate from. You need to fill in the ${DD_ARCHIVES_BUCKET}
  ## and ${DD_ARCHIVES_REGION} parameters to fit your S3 configuration.
  datadog_archives:
    type: datadog_archives
    inputs:
      - logs_finish_ddtags
    service: aws_s3
    bucket: ${DD_ARCHIVES_BUCKET}
    aws_s3:
      storage_class: "STANDARD"
      region: "${DD_ARCHIVES_REGION}"
EOT
}
```
{{% /tab %}}
{{< /tabs >}}

### Load balancing

{{< tabs >}}
{{% tab "Docker" %}}
Production-oriented setup is not included in the Docker instructions. Instead, refer to your company's standards for load balancing in containerized environments. If you are testing on your local machine, configuring a load balancer is unnecessary.
{{% /tab %}}
{{% tab "AWS EKS" %}}
Use the load balancers provided by your cloud provider.

> They adjust based on autoscaling events that the default Helm setup is configured for. The load balancers are internal-facing,

Reviewer 2: What is `they` referencing, the load balancers or the cloud provider?

Author: Load balancer, updated it.

NLBs provisioned by the [AWS Load Balancer Controller][1] are used.

See [Capacity Planning and Scaling][2] for load balancer recommendations when scaling the Worker.
#### Cross-availability-zone load balancing
The provided Helm configuration tries to simplify load balancing, but you must take into consideration the potential price implications of cross-AZ traffic. Wherever possible, the samples try to avoid creating situations where multiple cross-AZ hops can happen.

The sample configurations do not enable the cross-zone load balancing feature available in this controller. To enable it, add the following annotation to the `service` block:

```
service.beta.kubernetes.io/aws-load-balancer-attributes: load_balancing.cross_zone.enabled=true
```

See [AWS Load Balancer Controller][3] for more details.

[1]: https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/
[2]: /observability_pipelines/architecture/capacity_planning_scaling/
[3]: https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.4/guide/service/annotations/#load-balancer-attributes
{{% /tab %}}

{{% tab "APT-based Linux" %}}

> No built-in support for load-balancing is provided, given the single-machine nature of the installation. You will need to provision your own load balancers using whatever your company's standard is.

Reviewer 2: Optional:
```suggestion
Given the single-machine nature of the installation, there is no built-in support for load-balancing. Provision your own load balancers using your company's standard.
```

{{% /tab %}}
{{% tab "RPM-based Linux" %}}

> No built-in support for load-balancing is provided, given the single-machine nature of the installation. You will need to provision your own load balancers using whatever your company's standard is.

Reviewer 2: Optional:
```suggestion
Given the single-machine nature of the installation, there is no built-in support for load-balancing. You can provision your own load balancers with your company's standard.
```

{{% /tab %}}
{{% tab "Terraform (AWS)" %}}

> An NLB is provisioned by the Terraform module, and provisioned to point at the instances. Its DNS address is returned in the `lb-dns` output in Terraform.

Reviewer 2: I'm not sure if this reword makes more or less sense, but tried to make it in the active voice:
```suggestion
The Terraform module provisions an NLB to point at the instances. The DNS address is returned in the `lb-dns` output in Terraform.
```

{{% /tab %}}
{{< /tabs >}}

### Buffering

> Observability Pipelines includes multiple buffering strategies that allow you to increase the resilience of your cluster to downstream faults. The provided sample configurations use disk buffers, the capacities of which are rated for approximately 10 minutes of data at 10Mbps/core for Observability Pipelines deployments. That is often enough time for transient issues to resolve themselves, or for incident responders to decide what needs to be done with the observability data.

Reviewer 2: ```suggestion
Observability Pipelines includes multiple buffering strategies that allow you to increase the resilience of your cluster to downstream faults. The provided sample configurations use disk buffers, the capacities are rated for approximately 10 minutes of data at 10Mbps/core for Observability Pipelines deployments. That is often enough time for transient issues to resolve themselves, or for incident responders to decide what needs to be done with the observability data.
```

---

> title: Set Up Observability Pipelines to Send Logs in Datadog-Rehydratable Format to Amazon S3 and Datadog

Reviewer 2: Optional suggestion:
```suggestion
title: Send Logs in Datadog-Rehydratable Format to Amazon S3 and Datadog
```
